From 243afe20b1f7aa2f81fdb3646f1ec0a10de79073 Mon Sep 17 00:00:00 2001
From: Shayne Chen <shayne.chen@mediatek.com>
Date: Mon, 26 Aug 2024 22:38:37 +0800
Subject: [PATCH 126/149] mtk: hostapd: fix per-sta profile and bw indication
 element after upgrade

- Fix per-sta profile of ML probe resp during channel switch
- Align non-inheritance element to previous version
- Fix IE length of BW indication element

Signed-off-by: Shayne Chen <shayne.chen@mediatek.com>
---
 src/ap/beacon.c     | 6 +++---
 src/ap/beacon.h     | 1 +
 src/ap/hostapd.c    | 7 +++++++
 src/ap/ieee802_11.c | 2 +-
 4 files changed, 12 insertions(+), 4 deletions(-)

diff --git a/src/ap/beacon.c b/src/ap/beacon.c
index 892ebfeb7..0bd128efa 100644
--- a/src/ap/beacon.c
+++ b/src/ap/beacon.c
@@ -2962,8 +2962,8 @@ static size_t hostapd_add_sta_profile(struct ieee80211_mgmt *link_fdata,
 	bool ie_found;
 	u8 non_inherit_ele_ext_list[256] = { 0 };
 	u8 non_inherit_ele_ext_list_len = 0;
-	u8 non_inherit_ele_list[256] = { 0 };
-	u8 non_inherit_ele_list_len = 0;
+	u8 non_inherit_ele_list[256] = { WLAN_EID_VHT_CAP, WLAN_EID_VHT_OPERATION };
+	u8 non_inherit_ele_list_len = 2;
 	u8 num_link_elem_vendor_ies = 0, num_own_elem_vendor_ies = 0;
 	bool add_vendor_ies = false, is_identical_vendor_ies = true;
 	/* The bitmap of parsed EIDs. There are 256 EIDs and ext EIDs, so 32
@@ -3217,7 +3217,7 @@ static u8 * hostapd_gen_sta_profile(struct ieee80211_mgmt *link_data,
 }
 
 
-static void hostapd_gen_per_sta_profiles(struct hostapd_data *hapd)
+void hostapd_gen_per_sta_profiles(struct hostapd_data *hapd)
 {
 	bool tx_vap = hapd == hostapd_mbssid_get_tx_bss(hapd);
 	size_t link_data_len, sta_profile_len;
diff --git a/src/ap/beacon.h b/src/ap/beacon.h
index 809393902..a3c8d53a3 100644
--- a/src/ap/beacon.h
+++ b/src/ap/beacon.h
@@ -72,5 +72,6 @@ const u8 * hostapd_wpa_ie(struct hostapd_data *hapd, u8 eid);
 
 u8 * hostapd_unsol_bcast_probe_resp(struct hostapd_data *hapd,
 				    struct unsol_bcast_probe_resp *ubpr);
+void hostapd_gen_per_sta_profiles(struct hostapd_data *hapd);
 
 #endif /* BEACON_H */
diff --git a/src/ap/hostapd.c b/src/ap/hostapd.c
index 321d03434..7c76a03c2 100644
--- a/src/ap/hostapd.c
+++ b/src/ap/hostapd.c
@@ -4700,6 +4700,7 @@ void hostapd_chan_switch_config(struct hostapd_data *hapd,
 int hostapd_switch_channel(struct hostapd_data *hapd,
 			   struct csa_settings *settings)
 {
+	struct hostapd_data *link_bss;
 	int ret;
 
 	if (!(hapd->iface->drv_flags & WPA_DRIVER_FLAGS_AP_CSA)) {
@@ -4709,6 +4710,12 @@ int hostapd_switch_channel(struct hostapd_data *hapd,
 
 	ieee802_11_set_bss_critical_update(hapd, BSS_CRIT_UPDATE_EVENT_CSA);
 
+	if (hapd->conf->mld_ap) {
+		/* Generate per STA profiles for each affiliated APs */
+		for_each_mld_link(link_bss, hapd)
+			hostapd_gen_per_sta_profiles(link_bss);
+	}
+
 	ret = hostapd_fill_csa_settings(hapd, settings);
 	if (ret)
 		return ret;
diff --git a/src/ap/ieee802_11.c b/src/ap/ieee802_11.c
index f9b2086c1..f9ea37996 100644
--- a/src/ap/ieee802_11.c
+++ b/src/ap/ieee802_11.c
@@ -7501,7 +7501,7 @@ static u8 * hostapd_eid_bw_indication(struct hostapd_data *hapd, u8 *eid,
 {
 	u16 punct_bitmap = hostapd_get_punct_bitmap(hapd);
 	struct ieee80211_bw_ind_element *bw_ind_elem;
-	size_t elen = 3;
+	size_t elen = 4;
 
 	if (hapd->cs_freq_params.bandwidth <= 160 && !punct_bitmap)
 		return eid;
-- 
2.45.2

